modin/core/storage_formats/base/query_compiler.py

```diff
@@ -3232,6 +3232,37 @@ def repeat(self, repeats):
     index = property(_get_axis(0), _set_axis(0))
     columns = property(_get_axis(1), _set_axis(1))
 
+    def wrap_index(self, index):
```

Comment:
what do you think of this approach? this should be relatively safe and easy to maintain
```python
# base/query_compiler.py
def wrap_index(self, index):
    if hasattr(index, "__modin_qc_owner__"):
        return index
    else:
        setattr(index, "__modin_qc_owner__", self)
        return index

# modin/pandas/utils.py
def register_index_callback(attr_name, callback):
    add_callback_for_method_or_property_setter(target=pandas.Index, name=attr_name, callback=callback)
    add_callback_for_method_or_property_setter(target=pandas.RangeIndex, name=attr_name, callback=callback)
    add_callback_for_method_or_property_setter(target=pandas.MultiIndex, name=attr_name, callback=callback)
    ...

# modin/pandas/__init__.py
register_index_callback(attr_name="name", callback=propagate_new_name_to_qc)
register_index_callback(attr_name="names", callback=propagate_new_name_to_qc)
```

Reply:
Sorry, I didn't get the point. Do you propose to change pandas classes?

Reply:
yes, similar to the example we already have:
https://github.com/modin-project/modin/blob/c30ab4c132295b1e168fbb06a050c76554759e11/modin/pandas/utils.py#L384

I'm not sure though about potential consequences of this.

Reply:
That would be the easiest solution, but I don't know what could be the impact of it.

Reply:
@jbrockmendel any thoughts here?

Reply:
the callback approach looks simpler to me

Reply:
The only problem I see so far is using this kind of modified index inside partitions. As an Index object now has a link to the query compiler that contains modin frame that contains partitions, would this construction be reasonable or possible to serialize in order to transfer it to the plasma store?
```python

def kernel_function(df):
    hasattr(df.index, "__modin_qc_owner__") # will it be True or False?
    df.index.name = "new_name" # what will happen here?
    return df

print(hasattr(query_compiler.index, "__modin_qc_owner__")) # True
query_compiler._modin_frame._propagate_index_objs(axis=0) # Explicitly propagate the external index to partitions
query_compiler.apply_full_axis(axis=0, func=kernel_function)
```
We can probably remove the linkage to the query compiler when preparing the index for transferring, but this may be tricky in certain cases like when an index is passed not explicitly but implicitly via the function's scope:
```python
external_index = query_compiler.index
print(hasattr(external_index, "__modin_qc_owner__")) # True

def kernel_func(df):
    return df.index.equals(external_index)

query_compiler.apply_full_axis(axis=0, func=kernel_function)
```

Reply:
What mechanism of serialization is used there? For some methods, a wrapper is created, that substitutes the class with the original one, during the method invocation - https://github.com/modin-project/modin/pull/5152/files#diff-12b8f779be31989ef16320f2bc9386bbc1efa2963e29cac3706b3152d3812fc5R5210./

Here is an example of pickle serialization:
```python
df = pd.DataFrame({"A": [1, 2, 3]})
idx = df.index
print(f"Orig index: {type(idx)}")
print(f"Orig index qc: {hasattr(idx, '__modin_idx_qc__')}")
serialized = pickle.dumps(idx)
idx = pickle.loads(serialized)
print(f"Deserialized index: {type(idx)}")
print(f"Deserialized index qc: {hasattr(idx, '__modin_idx_qc__')}")
```
Output:
```
Orig index: <class 'modin.core.storage_formats.base.query_compiler.RangeIndex'>
Orig index qc: True
Deserialized index: <class 'pandas.core.indexes.range.RangeIndex'>
Deserialized index qc: False
```

Reply:
> What mechanism of serialization is used there?

It depends from engine to engine. For Ray, it's cloudpickle as far as I remember. Could you please try the same with `ray.get/ray.put`?

```python
idx = df.index
idx_serialized = ray.get(ray.put(idx))
print(hasattr(idx_serialized, "__modin_idx_qc__")) # ??
```

Reply:
Ray uses PandasQueryCompiler, for which this functionality is not enabled - https://github.com/modin-project/modin/pull/5152/files#diff-795c69dce7045f32efac397ed8a4559590e9b0a86f5b7dde015326b04f4e2a10R295 .

```python
df = pd.DataFrame({"A": [1, 2, 3]})
idx = df.index
print(type(df._query_compiler))
print(f"Orig index: {type(idx)}")
print(f"Orig index qc: {hasattr(idx, '__modin_idx_qc__')}")
df = ray.get(ray.put(idx))
print(f"Deserialized index: {type(idx)}")
print(f"Deserialized index qc: {hasattr(idx, '__modin_idx_qc__')}")
```
Output:
```
<class 'modin.core.storage_formats.pandas.query_compiler.PandasQueryCompiler'>
Orig index: <class 'pandas.core.indexes.range.RangeIndex'>
Orig index qc: False
Deserialized index: <class 'pandas.core.indexes.range.RangeIndex'>
Deserialized index qc: False
```

Reply:
I've enabled this functionality for Ray and now I see, that QC is not serialized:
```
<class 'modin.core.storage_formats.pandas.query_compiler.PandasQueryCompiler'>
Orig index: <class 'modin.core.storage_formats.base.query_compiler.RangeIndex'>
Orig index qc: True
Deserialized index: <class 'pandas.core.indexes.range.RangeIndex'>
Deserialized index qc: False
```

Reply:
I am still seeing that the functionality doesn't work for pandas storage format. Am I missing on something in testing?

```python
import modin.pandas as pd

df = pd.DataFrame({"a": [1, 2], "a": [3, 4]})
df.index.name = 'foo'
print(df.index)
# RangeIndex(start=0, stop=2, step=1, name='foo')
print(df._query_compiler._modin_frame._partitions[0][0].get().index)
RangeIndex(start=0, stop=2, step=1) # there is no name equal to foo
```

Reply:
This functionality is disabled for the pandas storage format.

Reply:
If the PR fixes the issue only for HDK, I propose to move `wrap_index` from base qc to hdk one.